### PR TITLE
Leverage Health Check Settings for non-container Services to Determine Need for Auto Redeployment

### PR DESCRIPTION
**Pull request by liyaqin1**
Related PRs: 
Telefig side: https://github.com/pinternal/teletraan-config-manager/pull/159 
UI side: https://github.com/pinterest/teletraan/pull/1476

### CODE REVIEW

**Comment by liyaqin1** [comment]
To evaluate this PR, I propose adding a health_check section to this configuration file: https://sourcegraph.pinadmin.com/github.com/pinternal/cosmos/-/blob/kvstore/scripts/teletraan/config/serviceset.conf. 
We can then run the test on a test/dev stage on Teletraan and closely monitor and observe the outcomes.
Screenshot 2024-02-29 at 4 29 56 PM

File: deploy-agent/deployd/common/utils.py [hunk 1]
```diff
@@ -267,7 +268,27 @@ def get_container_health_info(commit, service, redeploy) -> Optional[str]:
                                             tags={"status": "healthy", "service": service, "commit": commit})
             return returnValue
         else:
-            return None
+            fn = os.path.join("/mnt/deployd/", "{}_HEALTH_CHECK".format(service))
```
Comment (ntascii) [inline deploy-agent/deployd/common/utils.py:271]: can you create a helper fucntion for this code part?

File: deploy-agent/deployd/common/utils.py [hunk 2]
```diff
@@ -267,7 +268,27 @@ def get_container_health_info(commit, service, redeploy) -> Optional[str]:
...
+                healthcheckConfigs = dict((n.strip('\"\n\' ') for n in line.split("=", 1)) for line in f)
+                # The script configuration for auto-redeployment is currently activated for this non-container service
+                if "HEALTHCHECK_HTTP" not in healthcheckConfigs:
+                    return None
+                url = healthcheckConfigs["HEALTHCHECK_HTTP"]
+                if "http://" not in url:
+                    url = "http://" + url
+                resp = requests.get(url)
+                if resp.status_code >= 200 and resp.status_code < 300:
+                    return service + ":healthy"
```
Comment (ntascii) [inline deploy-agent/deployd/common/utils.py:285]: this can go to a healthcheck helper func as well.

File: deploy-agent/deployd/common/utils.py [hunk 3]
```diff
@@ -267,7 +268,27 @@ def get_container_health_info(commit, service, redeploy) -> Optional[str]:
                                             tags={"status": "healthy", "service": service, "commit": commit})
             return returnValue
         else:
-            return None
+            fn = os.path.join("/mnt/deployd/", "{}_HEALTH_CHECK".format(service))
+            if not os.path.isfile(fn):
+                return None
+        
+            with open(fn, 'r') as f:
+                healthcheckConfigs = dict((n.strip('\"\n\' ') for n in line.split("=", 1)) for line in f)
```
Comment (ntascii) [inline deploy-agent/deployd/common/utils.py:276]: why don't you store as a json object so you don't need to parse like this?

Comment (liyaqin1) [reply deploy-agent/deployd/common/utils.py:276]: This is quite straightforward and maintains the same format as the script configurations.

File: deploy-agent/deployd/common/utils.py (hunk 2 above)
Comment (liyaqin1) [reply deploy-agent/deployd/common/utils.py:285]: That sounds good. I'll proceed with updating the PR next Monday.

File: deploy-agent/deployd/common/utils.py (hunk 1 above)
Comment (liyaqin1) [reply deploy-agent/deployd/common/utils.py:271]: That sounds good. I'll proceed with updating the PR next Monday.

**Comment by liyaqin1** [comment]
Tested on devapp.
Test Case 1: Utilize a fully functional port for a service that is not housed within a container.
The config is:
```
yaqinli@dev-yaqinli:/mnt/deployd$ cat telefig_HEALTHCHECK 
HEALTHCHECK_HTTP=localhost:18132
yaqinli@dev-yaqinli:/mnt/deployd$ 
```
The output is:
```
2024-03-04 21:54:02,559 deployd:130 INFO the current service is: telefig
2024-03-04 21:54:02,559 deployd.common.utils:273 INFO Get health info for service telefig with commit caaae202a4baaaeea82449e20a330f30c0c11b05
2024-03-04 21:54:02,579 deployd.common.utils:249 INFO Get health info for service telefig with commit caaae202a4baaaeea82449e20a330f30c0c11b05
2024-03-04 21:54:02,580 deployd.common.utils:257 INFO Healthcheck is enabled on service telefig
2024-03-04 21:54:02,581 urllib3.connectionpool:250 DEBUG Starting new HTTP connection (1): localhost:18132
2024-03-04 21:54:02,582 urllib3.connectionpool:483 DEBUG http://localhost:18132 "GET / HTTP/1.1" 200 1705
2024-03-04 21:54:02,584 urllib3.connectionpool:1023 DEBUG Starting new HTTPS connection (1): statsboard.pinadmin.com:443
2024-03-04 21:54:02,599 urllib3.connectionpool:483 DEBUG https://statsboard.pinadmin.com:443 "PUT /api/v1/put/deployd.service_health_status?value=1&tags=status=healthy,service=telefig,commit=caaae202a4baaaeea82449e20a330f30c0c11b05,host=dev-yaqinli HTTP/1.1" 200 194
2024-03-04 21:54:02,599 deployd.common.stats:123 INFO Successfully send the metric to statsboard
```
The UI looks like:
Screenshot 2024-03-04 at 1 58 06 PM

Test Case 2: Utilize a non-functional port without enabling automatic redeployment for a non container service.
The config is:
```
yaqinli@dev-yaqinli:/mnt/deployd$ cat telefig_HEALTHCHECK 
HEALTHCHECK_HTTP=localhost:1111
yaqinli@dev-yaqinli:/mnt/deployd$ 
```
The output is:
```
2024-03-04 22:18:36,764 deployd:130 INFO the current service is: telefig
2024-03-04 22:18:36,764 deployd.common.utils:273 INFO Get health info for service telefig with commit caaae202a4baaaeea82449e20a330f30c0c11b05
2024-03-04 22:18:36,783 deployd.common.utils:249 INFO Get health info for service telefig with commit caaae202a4baaaeea82449e20a330f30c0c11b05
2024-03-04 22:18:36,783 deployd.common.utils:257 INFO Healthcheck is enabled on service telefig
2024-03-04 22:18:36,784 urllib3.connectionpool:250 DEBUG Starting new HTTP connection (1): localhost:1111
2024-03-04 22:18:36,786 urllib3.connectionpool:1023 DEBUG Starting new HTTPS connection (1): statsboard.pinadmin.com:443
2024-03-04 22:18:36,799 urllib3.connectionpool:483 DEBUG https://statsboard.pinadmin.com:443 "PUT /api/v1/put/deployd.service_health_status?value=1&tags=status=unhealthy,service=telefig,commit=caaae202a4baaaeea82449e20a330f30c0c11b05,host=dev-yaqinli HTTP/1.1" 200 196
2024-03-04 22:18:36,799 deployd.common.stats:123 INFO Successfully send the metric to statsboard
```
The UI is:
Screenshot 2024-03-04 at 2 18 57 PM

Test Case 3: Employ a non-functional port with automatic redeployment for a service not housed within a container.
the config is:
```
yaqinli@dev-yaqinli:/mnt/deployd$ cat telefig_HEALTHCHECK 
HEALTHCHECK_HTTP=localhost:1111
HEALTHCHECK_REDEPLOY_WHEN_UNHEALTHY=True
HEALTHCHECK_REDEPLOY_MAX_RETRY=2
yaqinli@dev-yaqinli:/mnt/deployd$ 
```
The output is:
```
2024-03-04 22:23:21,340 deployd:130 INFO the current service is: telefig
2024-03-04 22:23:21,341 deployd.common.utils:273 INFO Get health info for service telefig with commit caaae202a4baaaeea82449e20a330f30c0c11b05
2024-03-04 22:23:21,360 deployd.common.utils:249 INFO Get health info for service telefig with commit caaae202a4baaaeea82449e20a330f30c0c11b05
2024-03-04 22:23:21,360 deployd.common.utils:257 INFO Healthcheck is enabled on service telefig
2024-03-04 22:23:21,361 urllib3.connectionpool:250 DEBUG Starting new HTTP connection (1): localhost:1111
2024-03-04 22:23:21,362 deployd.common.utils:237 INFO Auto redeployment is enabled on service telefig
2024-03-04 22:23:21,362 deployd.common.utils:240 INFO current redeploy is 0
2024-03-04 22:23:21,363 urllib3.connectionpool:1023 DEBUG Starting new HTTPS connection (1): statsboard.pinadmin.com:443
2024-03-04 22:23:21,378 urllib3.connectionpool:483 DEBUG https://statsboard.pinadmin.com:443 "PUT /api/v1/put/deployd.service_health_status?value=1&tags=status=redeploy,service=telefig,commit=caaae202a4baaaeea82449e20a330f30c0c11b05,host=dev-yaqinli HTTP/1.1" 200 195
2024-03-04 22:23:21,378 deployd.common.stats:123 INFO Successfully send the metric to statsboard
```
And then get the output like this:
```
2024-03-04 22:27:40,935 deployd:130 INFO the current service is: telefig
2024-03-04 22:27:40,935 deployd.common.utils:273 INFO Get health info for service telefig with commit caaae202a4baaaeea82449e20a330f30c0c11b05
2024-03-04 22:27:40,953 deployd.common.utils:249 INFO Get health info for service telefig with commit caaae202a4baaaeea82449e20a330f30c0c11b05
2024-03-04 22:27:40,953 deployd.common.utils:257 INFO Healthcheck is enabled on service telefig
2024-03-04 22:27:40,955 urllib3.connectionpool:250 DEBUG Starting new HTTP connection (1): localhost:1111
2024-03-04 22:27:40,955 deployd.common.utils:237 INFO Auto redeployment is enabled on service telefig
2024-03-04 22:27:40,955 deployd.common.utils:240 INFO current redeploy is 1
2024-03-04 22:27:40,957 urllib3.connectionpool:1023 DEBUG Starting new HTTPS connection (1): statsboard.pinadmin.com:443
2024-03-04 22:27:40,970 urllib3.connectionpool:483 DEBUG https://statsboard.pinadmin.com:443 "PUT /api/v1/put/deployd.service_health_status?value=1&tags=status=redeploy,service=telefig,commit=caaae202a4baaaeea82449e20a330f30c0c11b05,host=dev-yaqinli HTTP/1.1" 200 195
2024-03-04 22:27:40,971 deployd.common.stats:123 INFO Successfully send the metric to statsboard
```
Eventually, the maximum retry limit was reached:
```
2024-03-04 22:29:05,038 deployd:130 INFO the current service is: telefig
2024-03-04 22:29:05,038 deployd.common.utils:273 INFO Get health info for service telefig with commit caaae202a4baaaeea82449e20a330f30c0c11b05
2024-03-04 22:29:05,057 deployd.common.utils:249 INFO Get health info for service telefig with commit caaae202a4baaaeea82449e20a330f30c0c11b05
2024-03-04 22:29:05,057 deployd.common.utils:257 INFO Healthcheck is enabled on service telefig
2024-03-04 22:29:05,058 urllib3.connectionpool:250 DEBUG Starting new HTTP connection (1): localhost:1111
2024-03-04 22:29:05,059 deployd.common.utils:237 INFO Auto redeployment is enabled on service telefig
2024-03-04 22:29:05,060 urllib3.connectionpool:1023 DEBUG Starting new HTTPS connection (1): statsboard.pinadmin.com:443
2024-03-04 22:29:05,076 urllib3.connectionpool:483 DEBUG https://statsboard.pinadmin.com:443 "PUT /api/v1/put/deployd.service_health_status?value=1&tags=status=unhealthy,service=telefig,commit=caaae202a4baaaeea82449e20a330f30c0c11b05,host=dev-yaqinli HTTP/1.1" 200 196
2024-03-04 22:29:05,076 deployd.common.stats:123 INFO Successfully send the metric to statsboard
```
The UI is:
Screenshot 2024-03-04 at 2 25 53 PM
and then :
Screenshot 2024-03-04 at 2 29 55 PM

Test Case 4: Employ a dysfunctional port with automatic redeployment, but without setting a maximum retry time for a non-container service.
The resultant output appears satisfactory and bears a close resemblance to Test Case 3. The sole disparity observed is the occurrence of one additional retry.

File: deploy-agent/deployd/common/utils.py [hunk 4]
```diff
@@ -231,9 +232,45 @@ def redeploy_check(labels, service, redeploy) -> int:
         return redeploy + 1
     return 0
     
+def redeploy_check_non_container(commit, service, redeploy, healthcheckConfigs):
+    if "HEALTHCHECK_REDEPLOY_WHEN_UNHEALTHY" in healthcheckConfigs and healthcheckConfigs["HEALTHCHECK_REDEPLOY_WHEN_UNHEALTHY"] == "True":
```
Comment (ntascii) [inline deploy-agent/deployd/common/utils.py:236]: if healthcheckConfigs.get("HEALTHCHECK_REDEPLOY_WHEN_UNHEALTHY") == "True":

File: deploy-agent/deployd/common/utils.py [hunk 5]
```diff
@@ -231,9 +232,45 @@ def redeploy_check(labels, service, redeploy) -> int:
...
+        log.info(f"Auto redeployment is enabled on service {service}")
+        if "HEALTHCHECK_REDEPLOY_MAX_RETRY" in healthcheckConfigs and redeploy < int(healthcheckConfigs["HEALTHCHECK_REDEPLOY_MAX_RETRY"]) or \
+            "HEALTHCHECK_REDEPLOY_MAX_RETRY" not in healthcheckConfigs and redeploy < 3:
```
Comment (ntascii) [inline deploy-agent/deployd/common/utils.py:239]: is 3 a default value? probably add a const some where for the default max retry redeploy?

File: deploy-agent/deployd/common/utils.py [hunk 6]
```diff
@@ -231,9 +232,45 @@ def redeploy_check(labels, service, redeploy) -> int:
...
+        if redeploy < max_retry:
+            log.info(f"redeploy is {redeploy}; max retry is {max_retry}")
+            send_statsboard_metric(name='deployd.service_health_status', value=1,
+                tags={"status": "redeploy", "service": service, "commit": commit})
+            return "redeploy-" + str(redeploy + 1)
+    send_statsboard_metric(name='deployd.service_health_status', value=1,
+        tags={"status": "unhealthy", "service": service, "commit": commit})
+    return service + ":unhealthy"
+
+def redeploy_check_for_none(commit, service, redeploy):
```
Comment (ntascii) [inline deploy-agent/deployd/common/utils.py:248]: can we have a better name? such as redeploy_check_without_containers?

**Review by ntascii — APPROVED** [review]

**Pull request merged**